### DISCUSSION TITLE
using symlinks as a doc-root is not atomic on apache per default

### DISCUSSION THREAD

**Discussion by staabm** [deployphp/deployer#2396] [Help needed]
when using deployer on a apache webserver (and therefore a symlinked docroot) is not atomic per default.

this means that requests might use release A when started but while apache is serving the request it might change to release B, because the symlink changed by a concurrently running deployment.

we see issues like Class-Not-Found, Class-Already-Declared, Cannot-redeclare-function because of this.

I found https://github.com/etsy/mod_realdoc on github which seems to patch this apache caveat.
see also this article from Rasmus Lerdorf on etsy: https://codeascraft.com/2013/07/01/atomic-deploys-at-etsy/

I don't know whether this also affects other webservers.

Is this something every knows and uses because this problem get not reported for this project more often?
Should it be documented somewhere, in case its 'required' (to prevent race conditions)?

**Comment by antonmedv** [comment]
I usually use nginx and making sure it properly configured. I use real path for this. 

I think it will be really good to document such thing in case in Apache.

**Reply by staabm** [reply]
We will try the mentioned apache module in production.

In case it works like expected, I will report back.

**Reply by Navi2016** [reply]
Hi,

Would love to know if the module worked for you?

**Reply by staabm** [reply]
we were not able to bring the module to production. therefore we sometime see deployment artifacts though (seem like race condition dependent).

sorry for not reporting back, as I initially promised.

**Reply by staabm** [reply]
in our future stack we use php-fpm, so the problem is kind of gone for us

**Reply by Navi2016** [reply]
Unfortunately it is not, we use Apache 2.4 and php-fpm and php-fpm requires a restart to see the new symlinks. Looking for a solution to that.

See: https://github.com/deployphp/deployer/discussions/3503
